src/utils/lazy.cpp

```diff
@@ -0,0 +1,113 @@
+#include "../../shared/utils/lazy.hpp"
+
+#include <cassert>
+#include <condition_variable>
+#include <mutex>
+
+constexpr std::uintptr_t UNINIT = 0x0;
+constexpr std::uintptr_t RUNNING = 0x1;
+constexpr std::uintptr_t INIT = 0x2;
+
+constexpr std::uintptr_t STATE_MASK = 0b11;
```

Comment:
A comment here would be nice to show the actual reason behind the bitmask here (and how it needs to fit with the others)